service/s3/s3crypto/s3crypto_client.go

```diff
@@ -0,0 +1,175 @@
+// do not load the entire contents into memory.
+type Client struct {
+	S3     *s3.S3
+	Config Config
+}
+
+// Config used to customize the Client
+type Config struct {
+	// Mode is used to dictate how we encrypt and decrypt the data.
+	//
+	// Defaults to AES GCM with the suggested pairing of AES Wrap/PRSA/KMS
+	Mode CryptoMode
+	// SaveStrategy will dictate where the envelope is saved.
+	//
+	// Defaults to the object's metadata
+	SaveStrategy
+	// InstructionSuffix is the instruction file name suffix. If it is empty, then
+	// the item key will be used followed by .instruction
+	InstructionSuffix string
+	// This is used to instantiate new kms clients when calling GetObject
+	KMSSession *session.Session
+	S3Session  *session.Session
+
+	// Used for instantiating non-kms key providers when getting objects
+	MasterKey []byte
+}
+
+// New instantiates a new S3 crypto client
+// Example:
+// cmkID := "some key id to kms"
+// kp, err = s3crypto.NewKMSKeyProvider(session.New(), cmkID, s3crypto.NewJSONMatDesc())
+// if err != nil {
+//   return err
+// }
+// svc := s3crypto.New(s3crypto.Authentication(kp))
+func New(mode CryptoMode, options ...func(*Client)) *Client {
+	sess := session.New()
+	client := &Client{
+		Config: Config{
+			Mode:         mode,
+			SaveStrategy: NewHeaderSaveStrategy(),
+			KMSSession:   sess,
+			S3Session:    sess,
+		},
+	}
+
+	for _, option := range options {
+		option(client)
+	}
+
+	client.S3 = s3.New(client.Config.S3Session)
+	return client
+}
+
+// PutObjectRequest creates a temp file to encrypt the contents into. It then streams
+// that data to S3.
+//
+// Example:
+// svc := s3crypto.New(s3crypto.Authentication(kp))
+// req, out := svc.PutObjectRequest(&s3.PutObjectInput {
+//   Key: aws.String("testKey"),
+//   Bucket: aws.String("testBucket"),
+//   Body: bytes.NewBuffer("test data"),
+// })
+// err := req.Send()
+func (c *Client) PutObjectRequest(input *s3.PutObjectInput) (*request.Request, *s3.PutObjectOutput) {
+	req, out := c.S3.PutObjectRequest(input)
+
+	// Create temp file to be used later for calculating the SHA256 header
+	f, err := ioutil.TempFile("./", "")
+	if err != nil {
+		req.Error = err
+		return req, out
+	}
+
+	req.Handlers.Build.PushFront(func(r *request.Request) {
+		md5 := newMD5Reader(input.Body)
+		sha := newSHA256Writer(f)
+		err = c.Config.Mode.EncryptContents(sha, md5)
+		if err != nil {
+			r.Error = err
+			return
+		}
+
+		shaHex := hex.EncodeToString(sha.GetValue())
+		req.HTTPRequest.Header.Set("X-Amz-Content-Sha256", shaHex)
+
+		f.Seek(0, 0)
+		input.Body = f
+		log.Println(input)
+
+		env, err := EncodeMeta(md5, c.Config.Mode)
+		if err != nil {
+			r.Error = err
+			return
+		}
+		err = c.Config.SaveStrategy.Save(env, input)
+		r.Error = err
+	})
+
+	fn := func(r *request.Request) {
+		// Close the temp file and cleanup
+		f.Close()
+		os.Remove(f.Name())
+	}
+	req.Handlers.Send.PushBack(fn)
+	req.Handlers.ValidateResponse.PushBack(fn)
```

Comment:
Whats the impact of doing the close and file delete in both send and ValidateResponse. Shouldn't it only be done once?